torchvision/transforms/_functional_tensor.py

```diff
@@ -300,23 +300,32 @@ def _rgb2hsv(img: Tensor) -> Tensor:
 
 def _hsv2rgb(img: Tensor) -> Tensor:
     h, s, v = img.unbind(dim=-3)
-    i = torch.floor(h * 6.0)
-    f = (h * 6.0) - i
+    h6 = h.mul(6)
+    i = torch.floor(h6)
+    f = h6.sub_(i)
     i = i.to(dtype=torch.int32)
 
-    p = torch.clamp((v * (1.0 - s)), 0.0, 1.0)
-    q = torch.clamp((v * (1.0 - s * f)), 0.0, 1.0)
-    t = torch.clamp((v * (1.0 - s * (1.0 - f))), 0.0, 1.0)
-    i = i % 6
+    sxf = s * f
+    one_minus_s = 1.0 - s
+    q = (1.0 - sxf).mul_(v).clamp_(0.0, 1.0)
+    t = sxf.add_(one_minus_s).mul_(v).clamp_(0.0, 1.0)
+    p = one_minus_s.mul_(v).clamp_(0.0, 1.0)
+    i.remainder_(6)
 
-    mask = i.unsqueeze(dim=-3) == torch.arange(6, device=i.device).view(-1, 1, 1)
+    vpqt = torch.stack((v, p, q, t), dim=-3)
 
-    a1 = torch.stack((v, q, p, p, t, v), dim=-3)
-    a2 = torch.stack((t, v, v, q, p, p), dim=-3)
-    a3 = torch.stack((p, p, t, v, v, q), dim=-3)
-    a4 = torch.stack((a1, a2, a3), dim=-4)
+    # vpqt -> rgb mapping based on i
+    select = torch.tensor([[0, 2, 1, 1, 3, 0], [3, 0, 0, 2, 1, 1], [1, 1, 3, 0, 0, 2]], dtype=torch.long)
 
-    return torch.einsum("...ijk, ...xijk -> ...xjk", mask.to(dtype=img.dtype), a4)
+    if img.device.type == "cuda":
+        select = select.pin_memory(img.device)
```

Comment:
Shouldn't this be 
```
    select = torch.tensor(
        [[0, 2, 1, 1, 3, 0], [3, 0, 0, 2, 1, 1], [1, 1, 3, 0, 0, 2]],
        dtype=torch.long
    ).to(device=img.device, non_blocking=True)
``` 
?

I see an improvement on 4D tensors only with this
```
[------------ HSV -> RGB cuda torch.float32 ------------]
                         |  fn_new2  |  fn_new  |  fn_v2 
1 threads: ----------------------------------------------
      (3, 400, 300)      |   259.1   |  426.3   |   275.9
      (3, 640, 768)      |   259.4   |  426.3   |   276.3
      (16, 3, 400, 300)  |   300.8   |  448.4   |  1157.4

Times are in microseconds (us).
```
vs previous results without `non_blocking=True`
```
[------------ HSV -> RGB cuda torch.float32 ------------]
                         |  fn_new2  |  fn_new  |  fn_v2 
1 threads: ----------------------------------------------
      (3, 400, 300)      |   260.9   |  428.9   |   274.1
      (3, 640, 768)      |   259.0   |  426.0   |   274.9
      (16, 3, 400, 300)  |   332.4   |  450.3   |  1157.1

Times are in microseconds (us).
```

Reply:
Absolutely, I'll update it tomorrow.